components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/PreIssueAccessTokenRequestBuilder.java

```diff
@@ -0,0 +1,330 @@
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AuthorizationGrantHandler;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+import org.wso2.carbon.identity.openidconnect.CustomClaimsCallbackHandler;
+import org.wso2.carbon.identity.openidconnect.OIDCClaimUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * This class is responsible for building the action execution request for the pre issue access token action.
+ */
+public class PreIssueAccessTokenRequestBuilder implements ActionExecutionRequestBuilder {
+
+    private static final Log LOG = LogFactory.getLog(PreIssueAccessTokenRequestBuilder.class);
+    public static final String CLAIMS_PATH_PREFIX = "/accessToken/claims/";
+    public static final String SCOPES_PATH_PREFIX = "/accessToken/scopes/";
+
+    @Override
+    public ActionType getSupportedActionType() {
+
+        return ActionType.PRE_ISSUE_ACCESS_TOKEN;
+    }
+
+    @Override
+    public ActionExecutionRequest buildActionExecutionRequest(ActionType actionType, Map<String, Object> eventContext)
+            throws ActionExecutionRequestBuilderException {
+
+        OAuthTokenReqMessageContext tokenMessageContext =
+                (OAuthTokenReqMessageContext) eventContext.get("tokenMessageContext");
+
+        Map<String, Object> additionalClaimsToAddToToken = getAdditionalClaimsToAddToToken(tokenMessageContext);
+
+        ActionExecutionRequest.Builder actionRequestBuilder = new ActionExecutionRequest.Builder();
+        actionRequestBuilder.actionType(actionType);
+        actionRequestBuilder.event(getEvent(tokenMessageContext, additionalClaimsToAddToToken));
+        actionRequestBuilder.allowedOperations(getAllowedOperations(additionalClaimsToAddToToken));
+
+        return actionRequestBuilder.build();
+    }
+
+    private Event getEvent(OAuthTokenReqMessageContext tokenMessageContext, Map<String, Object> claimsToAdd)
+            throws ActionExecutionRequestBuilderException {
+
+        OAuth2AccessTokenReqDTO tokenReqDTO = tokenMessageContext.getOauth2AccessTokenReqDTO();
+        AuthenticatedUser authorizedUser = tokenMessageContext.getAuthorizedUser();
+
+        PreIssueAccessTokenEvent.Builder eventBuilder = new PreIssueAccessTokenEvent.Builder();
+        eventBuilder.tenant(new Tenant(String.valueOf(IdentityTenantUtil.getTenantId(tokenReqDTO.getTenantDomain())),
+                tokenReqDTO.getTenantDomain()));
+
+        boolean isAuthorizedForUser = isAccessTokenAuthorizedForUser(tokenReqDTO.getGrantType(), tokenMessageContext);
+        if (isAuthorizedForUser) {
+            eventBuilder.organization(new Organization(authorizedUser.getUserResidentOrganization(),
```

Comment:
Why we are passing same arg two time

Reply:
This segment is yet to be improved considering b2b cases. I have forgotten to add the todo.
Will update